app/scripts/lib/multichain-api/wallet-requestPermissions.js

```diff
@@ -65,119 +68,125 @@ async function requestPermissionsImplementation(
   const [requestedPermissions] = params;
   delete requestedPermissions[Caip25EndowmentPermissionName];
 
-  const [_grantedPermissions] = await requestPermissionsForOrigin(
-    requestedPermissions,
-  );
-
-  // caveat values are frozen and must be cloned before modified
-  const grantedPermissions = { ..._grantedPermissions };
+  const legacyRequestedPermissions = pick(requestedPermissions, [
+    RestrictedMethods.eth_accounts,
+    PermissionNames.permittedChains,
+  ]);
+  delete requestedPermissions[RestrictedMethods.eth_accounts];
+  delete requestedPermissions[PermissionNames.permittedChains];
+
+  let legacyApproval;
+  if (Object.keys(legacyRequestedPermissions).length > 0) {
+    if (!legacyRequestedPermissions[RestrictedMethods.eth_accounts]) {
+      legacyRequestedPermissions[RestrictedMethods.eth_accounts] = {};
+    }
 
-  const ethAccountsPermission =
-    grantedPermissions[RestrictedMethods.eth_accounts];
+    if (!legacyRequestedPermissions[PermissionNames.permittedChains]) {
+      const { chainId } =
+        getNetworkConfigurationByNetworkClientId(networkClientId);
+      legacyRequestedPermissions[PermissionNames.permittedChains] = {
+        caveats: [
+          {
+            type: CaveatTypes.restrictNetworkSwitching,
+            value: [chainId],
+          },
+        ],
+      };
+    }
 
-  if (process.env.BARAD_DUR && ethAccountsPermission) {
-    // TODO: Use permittedChains permission returned from requestPermissionsForOrigin() when available
-    const { chainId } = getNetworkConfigurationByNetworkClientId(
-      req.networkClientId,
+    legacyApproval = await requestPermissionApprovalForOrigin(
```

Comment:
perhaps a comment explaining why this hook instead of `requestPermissionsForOrigin` -> we need to transform the permissions before granting them

Reply:
good idea. done here https://github.com/MetaMask/metamask-extension/pull/26483/commits/7e52c778a0a06f7abd6c82be1282a0ce9dc1df92